Add page explaining the image diff api (#1306)
chainguard-dev/edu: content/chainguard/chainguard-images/using-the-image-diff-api.md

---
title: "Using the Image Diff API"
type: "article"
description: "Learn how to use the Chainguard Images Diff API to compare changes between image versions."
date: 2023-12-10T08:49:31+00:00
lastmod: 2023-12-10T08:49:31+00:00
draft: false
tags: ["Chainguard Images", "Product", "API"]
images: []
menu:
  docs:
    parent: "chainguard-images"
weight: 450
toc: true
---

## Introduction to the Chainguard Images Diff API

> The Chainguard Images Diff API generates a JSON formatted list of package additions, removals, and detected CVEs in Chainguard Images. It is intended to provide an authenticated, machine readable, formatted list of changes between two image builds. Combined with the [Tag History API](/chainguard/chainguard-images/using-the-tag-history-api/), you can use the Diff API to compare arbitrary builds of an image over time and programmatically detect changes that might require gating a release or modifying your application code.

Reviewer: ```suggestion
The Chainguard Images Diff API generates a JSON-formatted list of package additions, removals, and detected CVEs in Chainguard Images. It is intended to provide an authenticated, machine readable, formatted list of changes between two image builds. Combined with the [Tag History API](/chainguard/chainguard-images/using-the-tag-history-api/), you can use the Diff API to compare arbitrary builds of an Image over time and programmatically detect changes that might require gating a release or modifying your application code.
```

> 

Reviewer: The diff API only works with Chainguard Images, right? I might make mention of that somewhere

Author: Right, added another Chainguard Images, thanks for the suggestion!

This guide will cover:

1. How to get an authentication token to query the Diff API

> 2. Identifying image digests in a repository for comparison

Reviewer: If the diff api only works with Chainguard Images, you might want to go through and change any `image` to `Image` to align with our product messaging

Author: Done, good call out

With these tools in place, you can begin gathering the information that you need to authenticate to and query the Diff API.

## Authenticating and Getting a Token

To get started, ensure you are authenticated to the Registry using the `chainctl auth login` command. Then locate your authentication token and create a shell variable for it.

If you are using Linux, run the following:

```sh
TOKEN=$(cat ~/.cache/chainguard/https\:--console-api.enforce.dev/oidc-token)
```

If you are using macOS, run the following:

```sh
TOKEN=$(cat ~/Library/Caches/chainguard/https:--console-api.enforce.dev/oidc-token)
```

Note: if at any point in this tutorial you receive an error like the following, your token has expired. Login and set your `$TOKEN` shell variable again:

```
{"code":16, "message":"failed to verify token", "details":[]}
```

For the purposes of this guide, we will use the `public-images` Chainguard IAM group. Set an environment variable corresponding to the group's [UIDP](/chainguard/administration/cloudevents/events-reference/#uidp-identifiers), which will be used later on:

```sh
GROUP_UIDP=$(chainctl iam group describe public-images -ojson |jq -r '.id')
```

> If you are using a private registry, be sure to substitute in your IAM group name in place of `public-images` to use your specific private group UIDP identifier in your subsequent commands.

Reviewer: ```suggestion
If you are using a private registry, be sure to substitute your IAM group name in place of `public-images` to use your specific private group UIDP identifier in your subsequent commands.
```

## Getting Image Digests to Compare

Now you will need to find image digests to compare. You can compare any two arbitrary digests using the API. In this example we will use the public `cgr.dev/chainguard/nginx` repository and find digests for the `latest` and `latest-dev` tags. You can also compare image digests for the same tag by combining data from the [Tag History API](/chainguard/chainguard-images/using-the-tag-history-api/).

The first digest in this example will be referred to as the `FROM_DIGEST`, and the second comparison image as `TO_DIGEST`.

Begin by creating a shell variable to hold the repository URL:

```sh
REPO_URL="cgr.dev/chainguard/nginx"
```

> Note: if you are using a private registry, replace `cgr.dev/chainguard/nginx` with your relevant private Chainguard Registry and repository name as required.

Reviewer: ```suggestion
If you are using a private registry, be sure to replace `cgr.dev/chainguard/nginx` with your relevant private Chainguard Registry and repository name as required.
```

I think it looks nicer if we only use "Note:" in places where we do a proper note block (with a `>` preceding it)

```sh
crane ls "${REPO_URL}" --omit-digest-tags
```

You will receive output like the following:

```
latest
latest-dev
next
```

Since we will compare the `latest` and `latest-dev` images in this guide, run the following to record their `digest` values as shell variables for reuse:

```sh
FROM_DIGEST=$(crane digest --platform "linux/amd64" "${REPO_URL}:latest")
TO_DIGEST=$(crane digest --platform "linux/amd64" "${REPO_URL}:latest-dev")
```

If you are using a different platform like `linux/arm64`, substitute it into the `crane` commands.

Now you can make a request to the Diff API using `curl` or programmatically with a language of your choice.

## Making a Request to the Diff API Using curl

To make API requests to the Diff API using `curl`, set a shell variable to the name of the image repository:

```sh
REPO_NAME=$(echo $REPO_URL | cut -d'/' -f3)
```

Now you can run the following command to record the repository UIDP identifier as a variable:

```sh
REPO_UIDP=$(curl -s -X GET -H "Authorization: Bearer $TOKEN" \
  "https://console-api.enforce.dev/registry/v1/repos?name=${REPO_NAME}" | \
  jq -r ".items[] | select(.id | startswith(\"${GROUP_UIDP}\")) | .id")
```

> Now with all of these variables set, you can call the Diff API with the following command:

Reviewer: ```suggestion
With all of these variables set, you can call the Diff API with the following command:
```

```sh
curl -s -X GET -H \
  "Authorization: Bearer $TOKEN" \
  "https://console-api.enforce.dev/registry/v1/repos/${REPO_UIDP}/diff?from_digest=${FROM_DIGEST}&to_digest=${TO_DIGEST}" | \
  jq -r
```

You will receive output like the following:

```
. . .
{
"packages": {
  "added": [
    . . .
    {
      "name": "bash",
      "version": "5.2.21-r0",
      "reference": "pkg:apk/wolfi/bash@5.2.21-r0?arch=x86_64"
    },
    {
      "name": "busybox",
      "version": "1.36.1-r2",
      "reference": "pkg:apk/wolfi/busybox@1.36.1-r2?arch=x86_64"
    },
    . . .
  "removed": [
    . . .
  ]
},
"vulnerabilities": {
  . . .
}
}
```

In this example output, the list of `added` packages include things like `bash` and other helpful utilities that are package in the nginx development image. If there are any vulnerabilities detected in an image they will be listed in that section of the output.

## Making a Request Programmatically to the Diff API Using Go

> If you would like to query the Diff API programmatically, there is an example Go application hosted in our [platform-examples](https://github.com/chainguard-dev/platform-examples/tree/main/image-diff) GitHub repository. The example uses the [Chainguard SDK](https://pkg.go.dev/chainguard.dev/sdk) and is a good way to learn how to interact with the API programmatically.

Reviewer: ```suggestion
If you would like to query the Diff API programmatically, there is an example Go application hosted in our [platform-examples](https://github.com/chainguard-dev/platform-examples/tree/main/image-diff) GitHub repository. The example uses the [Chainguard SDK](https://pkg.go.dev/chainguard.dev/sdk) and is a good way to learn how to interact with the API through a declarative approach.
```

I felt like "programmatically was feeling kind of repetitive

> Clone the repository, or copy the code manually and set some shell variables. Again we're using the `cgr.dev/chainguard/nginx` image as an example:

Reviewer: ```suggestion
Clone the repository, or copy the code manually. Again we're using the `cgr.dev/chainguard/nginx` image as an example:
```

```sh
git clone https://github.com/chainguard-dev/platform-examples
cd platform-examples/image-diff
```

> 

Reviewer: ```suggestion

Following that, set some shell variables:

```

Didn't make sense to me to have two separate code blocks like this
